### PR TITLE
gudhi pip package for version 3.2.0

### LINKED ISSUE #136: pip package
This isn't a priority since we have the conda-forge package. However, not everyone uses conda, and some of those users find gudhi hard to install. Having a pip package would, I believe, remove this excuse for not using gudhi :wink:.
This may not be easy, with the mix of cmake, c++, and dependencies to handle. But if someone wants to contribute such packaging, that would be great.

### PR DESCRIPTION
**Pull request by VincentRouvreau**
The result can be see here: https://test.pypi.org/project/gudhi/
You can install and test it by doing: pip install -i https://test.pypi.org/simple/ gudhi
Once the PR will be approved, I will build on pypi.org instead of test.pypi.org, and yml will be modified to build on release.
Be careful that a version cannot be undone !
Fix #136

### CODE REVIEW

File: .github/workflows/pip-packaging-windows.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
           architecture: x64
       - name: Install dependencies
         run: |
-          vcpkg install tbb:x64-windows boost-disjoint-sets:x64-windows boost-serialization:x64-windows boost-date-time:x64-windows boost-system:x64-windows boost-filesystem:x64-windows boost-units:x64-windows boost-thread:x64-windows boost-program-options:x64-windows eigen3:x64-windows mpfr:x64-windows mpir:x64-windows cgal:x64-windows
+          vcpkg install tbb:x64-windows boost-serialization:x64-windows boost-date-time:x64-windows boost-system:x64-windows boost-filesystem:x64-windows boost-units:x64-windows boost-thread:x64-windows boost-program-options:x64-windows eigen3:x64-windows mpfr:x64-windows mpir:x64-windows cgal:x64-windows
```
Comment (mglisse) [inline .github/workflows/pip-packaging-windows.yml:28]: Does boost-graph already come automatically as a dependency of some other boost thing?

**Comment by mglisse** [comment]
> Once the PR will be approved, I will build on pypi.org instead of test.pypi.org, and yml will be modified to build on release.

We were not doing releases in the github sense until now I think, so "release" here will really mean "build pip".

Would it make sense (sometimes?) to build a pip package from a branch, to make it easier for people to participate in the review? Of course it wouldn't be uploaded to pypi, just kept as an artifact (or at most uploaded to test.pypi).

> Be careful that a version cannot be undone !

But we can always do a 3.2.0.1 or 3.2.0.post1 if the upload of 3.2.0 went wrong. (by the way, according to PEP 440, it looks like you shouldn't put a dot before `rc` in 3.2.0.rc1.

File: .github/workflows/pip-packaging-linux.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+# on:
+#   release:
+#     types: [published]
+
+jobs:
+  build:
+    name: build pip wheels
+    runs-on: ubuntu-latest
+    container: gudhi/pip_for_gudhi
```
Comment (mglisse) [inline .github/workflows/pip-packaging-linux.yml:15]: It is a bit confusing to have ubuntu-latest immediately followed by a docker using some old linux, but if that's how it works...

**Comment by mglisse** [comment]
Are the logs of the builds available somewhere?

**Comment by VincentRouvreau** [comment]
> Are the logs of the builds available somewhere?

I don't know if it is available publicly, but it is in my github actions:
* Linux: https://github.com/VincentRouvreau/gudhi-devel/runs/727552932?check_suite_focus=true
* OSx: https://github.com/VincentRouvreau/gudhi-devel/runs/727552960?check_suite_focus=true
* Windows: https://github.com/VincentRouvreau/gudhi-devel/runs/727552981?check_suite_focus=true

In case it is not available publicly, I created this [gist](https://gist.github.com/VincentRouvreau/070960504b57ac253257ed723b85069c).

**Comment by VincentRouvreau** [comment]
> We were not doing releases in the github sense until now I think, so "release" here will really mean "build pip".

Yes that's it.

> Would it make sense (sometimes?) to build a pip package from a branch, to make it easier for people to participate in the review? Of course it wouldn't be uploaded to pypi, just kept as an artifact (or at most uploaded to test.pypi).

We can think about it. This means developers would require an account and a token on test.pypi and fill PYPI_PASSWORD in their accounts secrets. And how to increment versions ?

> But we can always do a 3.2.0.1 or 3.2.0.post1 if the upload of 3.2.0 went wrong. (by the way, according to PEP 440, it looks like you shouldn't put a dot before `rc` in 3.2.0.rc1.

Hmm... I don't think so, cf. https://www.python.org/dev/peps/pep-0440/#pre-releases
```
X.YaN   # Alpha release
X.YbN   # Beta release
X.YrcN  # Release Candidate
X.Y     # Final release
```
The same if we use `X.Y.Z`.

**Comment by mglisse** [comment]
> We can think about it. This means developers would require an account and a token on test.pypi and fill PYPI_PASSWORD in their accounts secrets.

Ah, yes. The plain artifact version may be simpler and just as useful, no need to upload it elsewhere as long as it is available somewhere.

> And how to increment versions ?

That's only needed if we upload. And then, for branches, it wouldn't have to increase, I would somehow use the commit hash as the last piece of the version number.

> X.YrcN  # Release Candidate

Currently in master we have `GUDHI_PATCH_VERSION 0.rc1` which should be `0rc1`, no?

> I don't know if it is available publicly

It is, thanks.

**Comment by VincentRouvreau** [comment]
> Currently in master we have `GUDHI_PATCH_VERSION 0.rc1` which should be `0rc1`, no?

Yes it should be 0rc1. If you look closer in the logs, you will see `make python` is complaining about this version number that does not respect pep440. My fault...

**Comment by mglisse** [comment]
> `make python` is complaining about this version number that does not respect pep440

That's exactly how I noticed it :wink:

**Comment by mglisse** [comment]
I tried a small benchmark of different builds with this simple test:
```python
import gudhi
import numpy as np
p=np.random.rand(30000,3)
d=gudhi.AlphaComplex(points=p).create_simplex_tree().persistence()
```
The debian package segfaults :cry: 
self build on debian (default flags): 5.4s
conda (with python 3.7): 5.1s
pip (with python 3.8): 5.6s

not quite as good as conda, but it isn't ridiculous, I mostly wanted to make sure that we were not accidentally building in debug mode or something like that.

File: src/python/setup.py.in [hunk 3]
```diff
@@ -62,13 +62,27 @@ for module in pybind11_modules:
...
+        'Documentation': 'https://gudhi.inria.fr/python/latest/',
+        'Source Code': 'https://github.com/GUDHI/gudhi-devel',
+        'License': 'https://gudhi.inria.fr/licensing/'
+        },
+    description='The Gudhi library is an open source library for ' \
+        'Computational Topology and Topological Data Analysis (TDA).',
+    long_description_content_type='text/x-rst',
+    long_description=long_description,
     ext_modules = ext_modules,
     install_requires = ['cython','numpy >= 1.9',],
```
Comment (mglisse) [inline src/python/setup.py.in:87]: We should remove cython from there, it isn't needed at runtime.

**Comment by mglisse** [comment]
Do you intend to "merge" this as a new branch (with 3.2 in its name) to be able to get a pip package for 3.2 without waiting for the next release?

**Comment by VincentRouvreau** [comment]
> Do you intend to "merge" this as a new branch (with 3.2 in its name) to be able to get a pip package for 3.2 without waiting for the next release?

Yes, I will generate gudhi pip package 3.2.0 from this branch on the official pypi.org website, then I will modify (`on release` instead of `on push`) it to generate automatically the pip package on each next gudhi releases. If you agree, and as tests seem ok, I can start to do it.

**Comment by mglisse** [comment]
ok. It isn't something we can test 100% beforehand anyway.

**Pull request merged**